src/template/oidrequest.tt

```diff
@@ -0,0 +1,483 @@
+             Source->DATA.METHOD_INFORMATION.OutputBufferLength;
+        Destination->DATA.METHOD_INFORMATION.BytesRead =
+             Source->DATA.METHOD_INFORMATION.BytesRead;
+        Destination->DATA.METHOD_INFORMATION.BytesNeeded =
+             Source->DATA.METHOD_INFORMATION.BytesNeeded;
+        Destination->DATA.METHOD_INFORMATION.BytesWritten =
+             Source->DATA.METHOD_INFORMATION.BytesWritten;
+        break;
+
+    case NdisRequestSetInformation:
+        Destination->DATA.SET_INFORMATION.BytesRead =
+             Source->DATA.SET_INFORMATION.BytesRead;
+        Destination->DATA.SET_INFORMATION.BytesNeeded =
+             Source->DATA.SET_INFORMATION.BytesNeeded;
+        break;
+
+    case NdisRequestQueryInformation:
+    case NdisRequestQueryStatistics:
+        Destination->DATA.QUERY_INFORMATION.BytesWritten =
+             Source->DATA.QUERY_INFORMATION.BytesWritten;
+        Destination->DATA.QUERY_INFORMATION.BytesNeeded =
+             Source->DATA.QUERY_INFORMATION.BytesNeeded;
+        break;
+    default:
+        break;
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+inline
+NDIS_STATUS
+NdisFPassthroughOidRequest(
+    _In_ NDIS_HANDLE NdisFilterHandle,
+    _In_ NDIS_OID_REQUEST *OidRequest,
+    _In_ ULONG PoolTag)
+/*++
+
+Routine Description:
+
+    Passes an unmodified OID request from the upper edge of a LWF to the LWF's
+    lower edge
+
+    Example usage:
+
+        NDIS_STATUS MyFilterOidRequest(
+            NDIS_HANDLE filterModuleContext,
+            NDIS_OID_REQUEST *oid)
+        {
+            MY_FILTER *filter = (MY_FILTER*)filterModuleContext;
+            return NdisFPassthroughOidRequest(filter->ndisHandle, oid, MY_TAG);
+        }
+
+    If you use NdisFPassthroughOidRequest, you should pair it with
+    NdisFPassthroughOidRequestComplete in the completion path.
+
+Arguments:
+
+    Destination - The OID request to receive the buffer length
+
+    Source - The OID request to provide the buffer length
+
+Return Value:
+
+    The return value is somewhat opaque to the caller, and should just be
+    returned back to NDIS from your FilterOidRequest handler.
+
+--*/
+{
+    NDIS_OID_REQUEST *Clone = NULL;
+
+    NDIS_STATUS NdisStatus = NdisAllocateCloneOidRequest(
+        NdisFilterHandle,
+        OidRequest,
+        PoolTag,
+        &Clone);
+    if (NDIS_STATUS_SUCCESS != NdisStatus)
+    {
+        NdisClearOidResult(OidRequest);
+        return NdisStatus;
+    }
+
+    NDIS_OID_REQUEST_SOURCE_RESERVED *Context =
+        (NDIS_OID_REQUEST_SOURCE_RESERVED*)&Clone->SourceReserved[0];
+    RtlZeroMemory(Context, sizeof(*Context));
+
+    Context->OriginalRequest = OidRequest;
+
+    Clone->RequestId = OidRequest->RequestId;
+
+    NdisStatus = NdisFOidRequest(NdisFilterHandle, Clone);
+    if (NDIS_STATUS_PENDING != NdisStatus)
+    {
+        NdisCopyOidResult(OidRequest, Clone);
+        NdisFreeCloneOidRequest(NdisFilterHandle, Clone);
```

Comment:
Won't this double free the clone OID request, since the completion handler would've been called?